src/dune_pkg/md5.ml

```diff
@@ -0,0 +1,32 @@
+open Stdune
+
+type t = string
+
+let equal = String.equal
+let to_hex_string = Stdlib.Digest.to_hex
+let to_dyn s = Dyn.variant "digest" [ String (to_hex_string s) ]
+
+let of_hex_string s =
+  match Stdlib.Digest.from_hex s with
+  | s -> Some s
+  | exception Invalid_argument _ -> None
+;;
+
+external md5_fd : Unix.file_descr -> string = "dune_pkg_md5_fd"
+
+let file file =
+  (* On Windows, if this function is invoked in a background thread, if can happen that
+     the file is not properly closed. [O_SHARE_DELETE] ensures that the main thread can
+     delete it even if it is still open. See #8243. *)
+  let file = Path.to_string file in
+  let fd =
+    match Unix.openfile file [ Unix.O_RDONLY; O_SHARE_DELETE; O_CLOEXEC ] 0 with
+    | fd -> fd
+    | exception Unix.Unix_error (Unix.EACCES, _, _) ->
+      raise (Sys_error (sprintf "%s: Permission denied" file))
```

Comment:
Don't think you need this conversion to `Sys_error`

Reply:
Should we raise a user error instead?

Reply:
Sure, not a bad idea.

Reply:
Although you might as well improve the error message for every error and not just `EACCES`